### ISSUE TITLE
Upload network to server?

### ISSUE THREAD

**Issue by Nyquixt** [leela-zero/leela-zero-server#222]
So, I tried to set up the LZ server locally. I managed to get everything done (I hope) except for the uploading network part. What I did was:
- gzip my network txt file (txt -> gz)
- change its name to `best-network.gz`
- use `curl -F 'weights=@best-network.gz' -F 'training_count=0' -F 'key=auth_key' http://localhost:8080/submit-network`
- copy `best-network.gz` into `leela-zero-server/network/` directory
- change `const Qstring server_url = "https://zero.sjeng.org/"` to `"http://localhost:8080/"`

...and still got this error about mismatching in hash.
BTW, is a code of 52 I got from `curl` bad?
Thanks in advance!
![curl52](https://user-images.githubusercontent.com/34199633/67623266-c3258480-f7e8-11e9-9b6d-25b796fc68bd.png)
![hashnomatch](https://user-images.githubusercontent.com/34199633/67623267-c4ef4800-f7e8-11e9-8071-671aa990bd8c.png)

**Comment by roy7** [comment]
Sounds similar to the problem in https://github.com/leela-zero/leela-zero-server/issues/217 but @Superkidd didn't actually say how he fixed it in the end.

**Comment by roy7** [comment]
BTW you don't need to rename to best-network.gz and it doesn't matter if you do. The server names each uploaded network with the hash, and stores it under networks/. For a net to become 'best' and promoted by server, you need to have it do a match and then the match successfully pass/promote the challenger. For the very first setup of the server you need to manually copy a network into networks/ and name it best-network.gz.

**Comment by Nyquixt** [comment]
@roy7 I did what you told me too. Yet, the calculated hash even with a different network uploaded is still 3ba11708... Also, I followed a more detailed instruction on how to set up the server on https://github.com/sai-dev/sai-server and still got the error. Is there anything else to change beside the QString url_server in autogtp? Plus, attached is the server log when I curled the network.
![Screenshot from 2019-10-28 16-00-55](https://user-images.githubusercontent.com/34199633/67717996-d711e780-f99c-11e9-9d83-d2a46cf5aeec.png)
![Screenshot from 2019-10-28 16-04-34](https://user-images.githubusercontent.com/34199633/67717998-dbd69b80-f99c-11e9-8313-9a48daad2a63.png)

**Comment by roy7** [comment]
Would you look at the download network with cat or a text editor and see if looks like a valid net? What's its file size? I'm guessing it isn't actually downloading a network so you don't see the hash changing. I'm suspecting it's downloading an error message or a null file or something?

The server error about trying to access 'count' of a null variable, is because you have no games at all. Once a game exists I guess that goes away. Try changing line 611 from "set.training_count = totalgames.count;" to "set.training_count = (totalgames ? totalgames.count || 0);" 

That should avoid crashing on upload of network with zero games played, and thus get the game into the database...

**Comment by roy7** [comment]
I've updated master branch to have everything in next branch, which could possibly have an effect. Some old next changes were related to gzip_hash I think.

**Comment by roy7** [comment]
@gcp When autogtp downloads a network, and it uses a URL like networks/hash.gz, we have a setting in the web server that intercepts the networks/ path don't we? Because I'm not seeing a handler anywhere in the server code to serve binary files requested under the networks/ end point. Doh?

**Comment by roy7** [comment]
This I think would be the key?

https://github.com/sai-dev/sai-server/commit/26c2ab81e44ebffd90b6c3a886727a688aa004cc

**Comment by roy7** [comment]
SAI team fixed this bug last year, who knew? https://github.com/sai-dev/sai-server/commit/02f7f102bd53d7db3eb6f94a014aafbb9c432eb6

**Comment by gcp** [comment]
>When autogtp downloads a network, and it uses a URL like networks/hash.gz, we have a setting in the web server that intercepts the networks/ path don't we?

Yes that's handled directly by the webserver without involving node.js. Improves the memory usage.

**Comment by roy7** [comment]
Thanks. I never documented that so it's why some people have failed to get the server working in the past, doh!

**Comment by gcp** [comment]
Currently there's even extra tricks to "fall back" to leela.online-go.com for older networks.

nginx config:
```
    location /networks {
        fancyindex on;
        fancyindex_default_sort date_desc;
        alias /home/jroy/code/leela-zero-server/network;
        try_files $uri @storage;
    }

    location @storage {
        expires 30s;
        return 301 https://leela.online-go.com$request_uri;
    }
```

**Comment by roy7** [comment]
Thanks, good to have that documented someplace. :) Sadly it's taken over a year for me to fix this by default for new people trying to run local servers.

**Comment by roy7** [comment]
Ok @Nyquixt check out https://github.com/leela-zero/leela-zero-server/pull/227

**Comment by Nyquixt** [comment]
@roy7 nice!

**Comment by Nyquixt** [comment]
BTW, why is 11 is on the second line of /best-nerwork-hash response?

**Comment by roy7** [comment]
Might be an old protocol version number or something? I'm not actually sure any more lol.
